src/term/transform/float_combinators.rs

```diff
@@ -1,277 +1,160 @@
     self.defs.extend(combinators);
   }
 }
 
-type Combinators = BTreeMap<Name, Definition>;
-
-#[derive(Debug)]
-struct TermInfo<'d> {
-  // Number of times a Term has been detached from the current Term
-  counter: u32,
-  def_name: Name,
-
-  builtin: bool,
-  needed_names: BTreeSet<Name>,
-  combinators: &'d mut Combinators,
-}
-
-impl<'d> TermInfo<'d> {
-  fn new(def_name: Name, builtin: bool, combinators: &'d mut Combinators) -> Self {
-    Self { counter: 0, def_name, builtin, needed_names: BTreeSet::new(), combinators }
-  }
-  fn request_name(&mut self, name: &Name) {
-    self.needed_names.insert(name.clone());
-  }
-
-  fn provide(&mut self, name: Option<&Name>) {
-    if let Some(name) = name {
-      self.needed_names.remove(name);
-    }
-  }
-
-  fn has_no_free_vars(&self) -> bool {
-    self.needed_names.is_empty()
-  }
-
-  fn replace_scope(&mut self, new_scope: BTreeSet<Name>) -> BTreeSet<Name> {
-    std::mem::replace(&mut self.needed_names, new_scope)
-  }
-
-  fn merge_scope(&mut self, target: BTreeSet<Name>) {
-    self.needed_names.extend(target);
-  }
-
-  fn detach_term(&mut self, term: &mut Term) {
-    let comb_name = Name::new(format!("{}$C{}", self.def_name, self.counter));
-    self.counter += 1;
-
-    let comb_var = Term::Ref { nam: comb_name.clone() };
-    let extracted_term = std::mem::replace(term, comb_var);
+impl Term {
+  fn float_combinators(
+    &mut self,
+    combinators: &mut Combinators,
+    name_gen: &mut usize,
+    book: &Book,
+    def_name: &Name,
+    builtin: bool,
+    seen: &mut IndexSet<Name>,
+  ) {
+    Term::recursive_call(move || {
+      for term in self.children_mut() {
+        // Don't float if it's safe or has unscoped variables.
+        if term.is_safe(book, seen) || term.has_unscoped_diff() {
+          continue;
+        }
 
-    let rules = vec![Rule { body: extracted_term, pats: Vec::new() }];
-    let rule = Definition { name: comb_name.clone(), rules, builtin: self.builtin };
-    self.combinators.insert(comb_name, rule);
-  }
-}
+        // Recursively float the children terms.
+        term.float_combinators(combinators, name_gen, book, def_name, builtin, seen);
 
-#[derive(Debug)]
-enum Detach {
-  /// Can be detached freely
-  Combinator,
-  /// Can not be detached
-  Unscoped { lams: BTreeSet<Name>, vars: BTreeSet<Name> },
-  /// Should be detached to make the program not hang
-  Recursive,
-}
+        match term {
+          // If it is an Application without free variables like '(bar 0)', float into a new definition.
+          Term::App { .. } => {
+            if term.free_vars().is_empty() && !term.has_unscoped_diff() {
+              float_combinator(def_name, name_gen, term, builtin, combinators);
+            }
+          }
 
-impl Detach {
-  fn can_detach(&self) -> bool {
-    !matches!(self, Detach::Unscoped { .. })
-  }
+          // If it is a Superposition, float every child element.
```

Comment:
We also don't need Sup as an expection. (by the way, this case is already covered by the chidren loop above)